### PR TITLE
Add Typescript type checking

### PR DESCRIPTION
**Pull request by shakyShane**
**Reviewer:** @GioSensation  

## Description

This PR enables Typescript with the most minimal changes I was able to use to get it passing.

Note: I have enabled some very strict settings, with more engineers working on this codebase now it should give us an element of safety.

We should look to incrementally add more and more types as each of us work on this codebase, and I am more than happy to help should anyone have any troubles.

### CODE REVIEW

**Comment by shakyShane** [comment]
@GioSensation Please note, `@ts-ignore` was used liberally at first to prevent this merging. It's better to type-check 95% of the codebase rather than none of it :)

**Comment by jonathanKingston** [comment]
> Hey there. I love this! ❤️💪 I will take a look tomorrow. My main concern right now is to avoid merge conflicts with @jonathanKingston 's branch. We might want to wait for his changes to land. Jonathan, what is your take?

I think like the test cases, there's more advantages to merging, it might actually catch some edge cases in my patches etc.

I think we should just go for it even if it's uncomfortable.

**Comment by shakyShane** [comment]
@jonathanKingston @GioSensation this is rebased and ready to merge 🥳

**Comment by GioSensation** [comment]
I'm reviewing now.

File: src/Form/Form.js [hunk 1]
```diff
@@ -233,7 +236,7 @@ class Form {
         if (input.nodeName !== 'SELECT') {
             const events = ['pointerdown']
             if (!isMobileApp) events.push('focus')
-            events.forEach((ev) => this.addListener(input, ev, handler, true))
```
Comment (GioSensation) [inline src/Form/Form.js:236]: Any reason to remove the capture flag? These usually ensure our listeners run before the page's own listeners.

File: src/Form/inputTypeConfig.js [hunk 2]
```diff
@@ -23,14 +23,15 @@ const getIdentitiesIcon = (input, {device}) => {
 
 /**
  * A map of config objects. These help by centralising here some complexity
- * @type {Object<SupportedMainTypes, InputTypeConfig>}
+ * TODO: Why is `emailNew` not part of 'SupportedMainTypes'?
```
Comment (GioSensation) [inline src/Form/inputTypeConfig.js:26]: We're removing `emailNew` everywhere. The new thing is `identities.emailAddress`. We still have to backport this properly to other platforms.

File: src/Form/inputTypeConfig.js [hunk 3]
```diff
@@ -23,14 +23,15 @@ const getIdentitiesIcon = (input, {device}) => {
...
+ * @type {Record<SupportedMainTypes & {'emailNew': InputTypeConfig}, InputTypeConfig>}
  */
 const inputTypeConfig = {
     emailNew: {
```
Comment (GioSensation) [inline src/Form/inputTypeConfig.js:30]: I think I had already removed this. Maybe I remember incorrectly. Anyway, this should be removed soon.

File: src/UI/Tooltip.js [hunk 4]
```diff
@@ -70,16 +79,18 @@ class Tooltip {
...
         this.input = input
         this.associatedForm = associatedForm
         this.interface = Interface
+        this.count = 0
     }
     append () {
         document.body.appendChild(this.host)
     }
     remove () {
-        window.removeEventListener('scroll', this.checkPosition, {passive: true, capture: true})
```
Comment (GioSensation) [inline src/UI/Tooltip.js:82]: I see why we removed this (["Note: The basic scroll event cannot be canceled"](https://developers.google.com/web/updates/2016/06/passive-event-listeners?hl=en)), but we should also remove it from where we add this listener. Do you mind double checking?

File: src/autofill-utils.js [hunk 5]
```diff
@@ -156,15 +157,19 @@ const safeExecute = (el, fn) => {
     const intObs = new IntersectionObserver((changes) => {
         for (const change of changes) {
             // Feature detection
+            // @ts-ignore todo: Why does TS not understand .isVisible here?
```
Comment (GioSensation) [inline src/autofill-utils.js:160]: It's a new version of the API only available in Chrome: https://web.dev/intersectionobserver-v2/.

File: src/autofill-utils.js [hunk 6]
```diff
@@ -156,15 +157,19 @@ const safeExecute = (el, fn) => {
...
+                // @ts-ignore
                 change.isVisible = true
             }
+            // @ts-ignore
             if (change.isIntersecting && change.isVisible) {
                 fn()
             }
         }
         intObs.disconnect()
+    // @ts-ignore todo: Why does TD think trackVisibility is invalid?
```
Comment (GioSensation) [inline src/autofill-utils.js:172]: See previous comment.

File: src/device-interface.d.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+     credentials: CredentialsObject[],
+     creditCards: CreditCardObject[],
+     identities: IdentityObject[],
+}
+
+type APIResponse<Type> = Promise<{ success: [Type], error?: string }>
+
+interface Window {
+     EmailInterface: any
+     webkit: any
```
Comment (GioSensation) [inline src/device-interface.d.ts:49]:
```suggestion
     // Used in the Android app
     EmailInterface: any
     // Used in Apple apps
     webkit: any
```

**Review by GioSensation — COMMENTED** [review]
A few comments to look into, but overall this is great. Thanks, Shane!

File: src/Form/Form.js (hunk 1 above)
Comment (shakyShane) [reply src/Form/Form.js:236]: yeah, because `this.addListener` doesn't accept a 4th param

https://github.com/duckduckgo/duckduckgo-autofill/blob/91229cef823fcd75b1fcc03eae51414ae9a385f9/src/Form/Form.js#L186-L189

File: src/Form/inputTypeConfig.js (hunk 2 above)
Comment (shakyShane) [reply src/Form/inputTypeConfig.js:26]: ok, changing the comment to reflect this

File: src/UI/Tooltip.js (hunk 4 above)
Comment (shakyShane) [reply src/UI/Tooltip.js:82]: Yep, I'll remove it. Actually I didn't remove it for that reason, I removed it because TS told me that option was not available when 'removing' an event :)

File: src/autofill-utils.js (hunk 5 above)
Comment (shakyShane) [reply src/autofill-utils.js:160]: I'll update the comment

File: src/Form/Form.js (hunk 1 above)
Comment (GioSensation) [reply src/Form/Form.js:236]: Ah, yeah, of course. Brainfart.

File: src/autofill-utils.js (hunk 5 above)
Comment (shakyShane) [reply src/autofill-utils.js:160]: Actually, we can add these 🥳

**Review by GioSensation — APPROVED** [review]
LGTM. 🚀
